dali/pipeline/operators/reader/reader_op.h

```diff
@@ -228,7 +241,28 @@ class DataReader : public Operator<Backend> {
     return loader_->Size();
   }
 
+
+  LoadTarget* GetSample(int sample_idx) {
+    return prefetched_batch_queue_[curr_batch_consumer_][sample_idx];
+  }
+
  protected:
+  int NextIdx(int curr_batch, bool& cycle) {
```

Comment:
I would create SetFull and SetEmpty functions so it is easy to track that IsEmpty/SetEmpty and IsFull/SetFull need to be protected by the same mutex. Now it is hard to track this and probably we check IsEmpty in the mutex while modify values inside it without mutex.

Reply:
Maybe instead of returning value this function can accept third argument and modify it as well. Now one value is returned and second modified as an argument.

Reply:
@JanuszL We are not explicitely saying when it's full and when it's empty. We are just moving the pointers to the current index for the producer and consumer and based on that we determine whether we are full or empty.
I agree with having some small struct/members to handle the advances of those pointers in a more convenient way

Reply:
Still some comment that NextIdx can trigger wake up will be good. Also moving `producer_.notify_one();` to NextIdx could better show what NextIdx is suppose to do and how it can affect producer_

Reply:
I did a small refactoring, I believe the code is more readable now. Please take a look